src/System.Net.NameResolution/src/System/Net/DNS.cs

```diff
@@ -0,0 +1,383 @@
+            IAsyncResult asyncResult = HostResolutionBeginHelper(hostNameOrAddress, false, requestCallback, stateObject);
+
+            if (Logging.On) Logging.Exit(Logging.Sockets, "DNS", "BeginGetHostEntry", asyncResult);
+            return asyncResult;
+        } // BeginResolve
+
+        public static IAsyncResult BeginGetHostEntry(IPAddress address, AsyncCallback requestCallback, object stateObject)
+        {
+            if (Logging.On) Logging.Enter(Logging.Sockets, "DNS", "BeginGetHostEntry", address);
+            IAsyncResult asyncResult = HostResolutionBeginHelper(address, true, true, requestCallback, stateObject);
+
+            if (Logging.On) Logging.Exit(Logging.Sockets, "DNS", "BeginGetHostEntry", asyncResult);
+            return asyncResult;
+        } // BeginResolve
+
+        public static IPHostEntry EndGetHostEntry(IAsyncResult asyncResult)
+        {
+            if (Logging.On) Logging.Enter(Logging.Sockets, "DNS", "EndGetHostEntry", asyncResult);
+            IPHostEntry ipHostEntry = HostResolutionEndHelper(asyncResult);
+
+            if (Logging.On) Logging.Exit(Logging.Sockets, "DNS", "EndGetHostEntry", ipHostEntry);
+            return ipHostEntry;
+        } // EndResolve()
+
+        public static IAsyncResult BeginGetHostAddresses(string hostNameOrAddress, AsyncCallback requestCallback, object state)
+        {
+            if (Logging.On) Logging.Enter(Logging.Sockets, "DNS", "BeginGetHostAddresses", hostNameOrAddress);
+            IAsyncResult asyncResult = HostResolutionBeginHelper(hostNameOrAddress, true, requestCallback, state);
+
+            if (Logging.On) Logging.Exit(Logging.Sockets, "DNS", "BeginGetHostAddresses", asyncResult);
+            return asyncResult;
+        } // BeginResolve
+
+        public static IPAddress[] EndGetHostAddresses(IAsyncResult asyncResult)
+        {
+            if (Logging.On) Logging.Enter(Logging.Sockets, "DNS", "EndGetHostAddresses", asyncResult);
+            IPHostEntry ipHostEntry = HostResolutionEndHelper(asyncResult);
+
+            if (Logging.On) Logging.Exit(Logging.Sockets, "DNS", "EndGetHostAddresses", ipHostEntry);
+            return ipHostEntry.AddressList;
+        } // EndResolveToAddresses
+
+        //************* Task-based async public methods *************************
+        public static Task<IPAddress[]> GetHostAddressesAsync(string hostNameOrAddress)
+        {
+            return Task<IPAddress[]>.Factory.FromAsync(BeginGetHostAddresses, EndGetHostAddresses, hostNameOrAddress, null);
+        }
+
+        public static Task<IPHostEntry> GetHostEntryAsync(IPAddress address)
+        {
+            return Task<IPHostEntry>.Factory.FromAsync(BeginGetHostEntry, EndGetHostEntry, address, null);
+        }
+
+        public static Task<IPHostEntry> GetHostEntryAsync(string hostNameOrAddress)
+        {
+            return Task<IPHostEntry>.Factory.FromAsync(BeginGetHostEntry, EndGetHostEntry, hostNameOrAddress, null);
+        }
```

Comment:
The nice thing about FromAsync is that it makes these easy to do as one-liners.  The bad thing about FromAsync is that there are some hidden costs here, e.g. the delegate allocations.  Not for this PR, but we shoudl follow-up on these and any others to ensure we're doing them in an optimal way.